### ISSUE TITLE
[ENH] design - 2nd party package integration, decentral estimator library

### ISSUE THREAD

**Issue by fkiraly** [sktime/sktime#6639]
Opening a discussion on 2nd party integration patterns and on how (or whether) to build a decentral estimator library.

More concretely, what are the base patterns in setting up packages like `prophetverse` or `skchoice` that provide `sktime` compatible estimators to direct users, following the unified `sktime` API?

FYI @felipeangelimvieira, @tveten

Some starter thoughts of mine about requirements and principles.

Operational, governance desiderata:

* decentral - 2nd party maintainers can manage their package freely
* gatekeeperless - not one person or small group blocking/approving
* indexed - users can easily find and search, using types and tags
* simple 2nd party developer setup - like publishing on pypi
* maximizing credit - visible contributor and owner assignment

Technical desiderata:

* unified API - all estimators follow strongly compatible interfaces and types
* easy API conformance checking
* queriable estimator index - there probably needs to be a central index
* estimator level dependency management
* minimizing integration complexity

Short-term, how would this look like?

* unified API and tests defined in `sktime`
* estimators *are* mini-packages with own dependencies - similar to current `sktime`
* use of `sktime` proper as a package index - docstring plus tags
* 2nd party package runs `check_estimator`

Questions I have:

* pattern "use import from soft dependency if available, use docstring/tag identical copy if not" - any problems with using different classes depending on environment?
* where should tests live? Certainly in the 2nd party package. Any testing in `sktime`? Does not scale if "test all"; no clear mechanism to trigger on change otherwise?
* how to minimize complexity of the developer experience?
* how to ensure that credit is assigned to the 2nd party package and its developers?

**Comment by felipeangelimvieira** [comment]
Franz, I have a few additional questions/points to consider:

- 2nd party packages, such as Prophetverse, may have specific methods (such as [predict_all_samples](https://github.com/felipeangelimvieira/prophetverse/blob/aa55b59c83e38998229c42025acb951f27c41637/src/prophetverse/sktime/base.py#L225), which returns all posterior samples from the model). Would that create any compatibility issues with sktime?
- Some arguments in Prophetverse models are custom objects, such as those passed in exogenous_effects [(docs here)](https://felipeangelimvieira.github.io/prophetverse/reference/sktime/prophet/). These objects have their own inner hyperparameters. Should they implement get_params() behavior so that their parameters can be set with set_params()? Or would that be ideal but not essential?

**Comment by fkiraly** [comment]
Linking discussion about dependency handling and update/deprecations workflows here:

https://github.com/felipeangelimvieira/prophetverse/issues/64#issuecomment-2178550987

**Comment by fkiraly** [comment]
On discord, @yarnabrina highlighted the following problem:

Suppose we have a 2nd party package like `skchange`, where `skchange` imports `sktime` for `BaseEstimator` or similar. That import is non-negotiable because the package uses `sktime` base classes.

What is the policy on dependencies in `sktime`? In this special case, we import `skchange` in the notebook, but should it also be imported, say, in the `all_extras` and `detection` dependency sets?

Circular package level dependencies could be problematic, and if it were not for the notebook, they could be avoided by using the placeholder pattern for indexing, as used for instance with `prophetverse`.

**Comment by fkiraly** [comment]
FYI @tveten, @felipeangelimvieira, as you are directly impacted by the above observation.

**Comment by felipeangelimvieira** [comment]
Thank you Franz. In that case, Prophetverse is already safe with respect to circular dependencies, as you said right? One interesting pattern to look at is the one in Airflow ([Airflow Providers](https://airflow.apache.org/docs/apache-airflow-providers/index.html))

**Comment by fkiraly** [comment]
> In that case, Prophetverse is already safe with respect to circular dependencies, as you said right?

Yes, `sktime` does not require `prophetverse` in its soft dependencies, it only checks if present in the placeholder import. This is "safe" as the tests are in `prophetverse`, which depends on `sktime` for the interface and the `parametrize_with_checks` tests.

But now suppose we wanted to use the "effects" base classes in another model, or do sth similar, that would create a problem.

**Comment by fkiraly** [comment]
> One interesting pattern to look at is the one in Airflow ([Airflow Providers](https://airflow.apache.org/docs/apache-airflow-providers/index.html))

Do you know how this works under the hood, @felipeangelimvieira? Is it the same pattern that the providers import the core, and the core checks for presence?
